src/components/swap/SwapLineItem.tsx

```diff
@@ -0,0 +1,250 @@
+        any of these fees.
+      </Trans>{' '}
+      <ExternalLink href="https://support.uniswap.org/hc/en-us/articles/18673568523789-What-is-a-token-fee-">
+        Learn more
+      </ExternalLink>
+    </>
+  )
+}
+
+function Loading({ width = 50 }: { width?: number }) {
+  return <LoadingRow data-testid="loading-row" height={15} width={width} />
+}
+
+function ExchangeRateRow({ trade }: { trade: InterfaceTrade }) {
+  const { formatNumber } = useFormatter()
+  const rate = `1 ${trade.executionPrice.quoteCurrency.symbol} = ${formatNumber({
+    input: parseFloat(trade.executionPrice.toFixed(9)),
+    type: NumberType.TokenTx,
+  })} ${trade.executionPrice.baseCurrency.symbol}`
+  return <>{rate}</>
+}
+
+function ColoredPercentRow({ percent }: { percent: Percent }) {
+  const { formatPriceImpact } = useFormatter()
+  return <ColorWrapper textColor={getPriceImpactColor(percent)}>{formatPriceImpact(percent)}</ColorWrapper>
+}
+
+function CurrencyAmountRow({ amount }: { amount: CurrencyAmount<Currency> }) {
+  const { formatCurrencyAmount } = useFormatter()
+  const formattedAmount = formatCurrencyAmount({ amount, type: NumberType.SwapDetailsAmount })
+  return <>{`${formattedAmount} ${amount.currency.symbol}`}</>
+}
+
+type LineItemData = {
+  Label: React.FC
+  Value: React.FC
+  TooltipBody?: React.FC
+  tooltipSize?: TooltipSize
+  loaderWidth?: number
+}
+
+function useLineItem(props: SwapLineItemProps): LineItemData | undefined {
+  const { trade, syncing, allowedSlippage, type } = props
+  const { formatNumber } = useFormatter()
+
+  const isUniswapX = isUniswapXTrade(trade)
+  const isPreview = isPreviewTrade(trade)
+  const chainId = trade.inputAmount.currency.chainId
+
+  // Used to hide price impact row if last trade was UniswapX
+  const [wasUniswapX, setWasUniswapX] = useState<boolean>(isUniswapX)
+  useEffect(() => setWasUniswapX((prev) => (!isPreview ? isUniswapX : prev)), [isUniswapX, isPreview])
+
+  switch (type) {
+    case SwapLineItemTypes.EXCHANGE_RATE:
+      return {
+        Label: () => <Trans>Exchange rate</Trans>,
+        Value: () => <ExchangeRateRow trade={trade} />,
+      }
+    case SwapLineItemTypes.NETWORK_FEE:
+      if (!SUPPORTED_GAS_ESTIMATE_CHAIN_IDS.includes(chainId)) return
+      return {
+        Label: () => <Plural value={getTransactionCount(trade) || 1} one="Network fee" other="Network fees" />,
+        TooltipBody: () => <GasBreakdownTooltip trade={trade} hideUniswapXDescription />,
+        Value: () => {
+          if (isPreview) return <Loading />
+          return <>{formatNumber({ input: trade.totalGasUseEstimateUSD, type: NumberType.FiatGasPrice })}</>
+        },
+      }
+    case SwapLineItemTypes.PRICE_IMPACT:
+      // Hides price impact row if the current trade is UniswapX or if we're previewing a trade when the last trade was UniswapX
+      if (isUniswapX || (wasUniswapX && isPreview)) return
```

Comment:
hmm is the rationale here that if the previous trade was uniswapx, and we're currently in a preview trade, that the next trade will also probably be uniswapx? bit unclear rn why the hide logic is `(isUniswapX || (wasUniswapX && isPreview))`

Reply:
Yes the goal is to predict the trade so that the UI is as non-jumpy as possible! I'll tweak the naming/commenting to make it clearer

you can see what I'm going for in the videos below:

prod:

https://github.com/Uniswap/interface/assets/39385577/1c4599f6-ffa9-464d-b299-cf0d9dc14228

preview:

https://github.com/Uniswap/interface/assets/39385577/1c4c8cfd-dcd0-4be4-9899-88177a2d92ed

Reply:
ok nice!